src/query/storages/common/cache/src/providers/disk_cache.rs

```diff
@@ -132,11 +133,165 @@ where C: Cache<String, u64, DefaultHashBuilder, FileSize>
         Ok(self)
     }
 
+    fn fuzzy_restart(self) -> result::Result<Self> {
+        fs::create_dir_all(&self.root)?;
+
+        let parallel_degree = std::thread::available_parallelism()
+            .expect("failed to detect number of parallelism")
+            .get();
+
+        let root_dir = Path::new(&self.root);
+        assert!(root_dir.is_dir());
+
+        let mut sub_dirs = Vec::new();
+        for entry in fs::read_dir(root_dir)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_dir() {
+                sub_dirs.push(path);
+            } else {
+                warn!(
+                    "unexpected file {:?} found under cache dir {:?}",
+                    path, root_dir
+                );
+            }
+        }
+
+        info!("{} sub dirs found {:?}", sub_dirs.len(), sub_dirs);
+
+        if sub_dirs.is_empty() {
+            info!("no previous cache found");
+            return Ok(self);
+        }
+
+        // The smallest unit of parallelism is a subdirectory.
+        let parallel_degree = std::cmp::min(sub_dirs.len(), parallel_degree);
+
+        info!("launching {} threads to reload cache keys", parallel_degree);
+
+        // number of sub dirs should be limited, use unbounded channel
+        let (tx, rx) = crossbeam_channel::unbounded();
+
+        let mut handles = vec![];
+
+        let root = self.root.clone();
+        let cache_holder = Arc::new(RwLock::new(Some(self)));
+
+        for thread_num in 0..parallel_degree {
+            let cache_holder = cache_holder.clone();
+            let rx = rx.clone();
+            let root = root.clone();
+            let thread_builder =
+                std::thread::Builder::new().name("table-data-cache-fuzzy-restart".to_owned());
+            let handle = thread_builder.spawn(move || {
```

Comment:
Here why we not to use `try_spawn_batch_with_owned_semaphore` (or like) in the runtime? https://github.com/datafuselabs/databend/blob/main/src/common/base/src/runtime/runtime.rs#L251?

Reply:
semaphore is not needed here to control concurrency. it is a very simple job: launch n threads to process m directories, where  m >= n, let's keep it simple.